README.md
# React Facebook Login HOC
[![npm](https://img.shields.io/npm/v/react-facebook-login-hoc.svg?maxAge=2592000)](https://www.npmjs.com/package/react-facebook-login-hoc)

## What is ?

<!-- React Facebook Login Higher Order Component for you to easily to initialize and use relative method. -->
Easily for you to append facebook **root dom** and **dependencies file**

## Installation

```
npm install react-facebook-login-hoc
```

## Exposed API

* this.props.fb.status(callback(response): function) - callback function will response facebook login auth status

* this.props.fb.login(callback(response): function [,scope: object]) - callback function will response login result

* this.props.fb.logout()

## default Props

|    params    |   value  |             default value            |
|:------------:|:--------:|:------------------------------------:|
|     appId    |  string  |              Required                |
|     scope    |  string  |           public_profile             |
|     xfbml    |  boolean |                 false                |
|     cookie   |  boolean |                 false                |
|    version   |  string  |                  2.6                 |
|    language  |  string  |                  en_US               |

## Configuration

```javascript
const configureLoginProps = {
  appId: 'Your APP ID',
  scope: 'public_profile, email',
  xfbml: false,
  cookie: false,
  version: 2.6,
  language: 'en_US'
}
```

## How to use ?

```javascript
import React, { Component } from 'react'
import { render } from 'react-dom'

import LoginHOC from 'react-facebook-login-hoc'

const configureLoginProps = {
  scope: 'public_profile',
  xfbml: false,
  cookie: false,
  version: 2.6,
  language: 'en_US',
  appId: '488387194689361'
}

class App extends Component {
  constructor(props) {
    super(props)

    this.status = this.props.fb.status
    this.login = this.props.fb.login
    this.logout = this.props.fb.logout
  }
  getStatus(response) {
    if (response.authResponse) {
      this.responseApi.call(this, response.authResponse)
    }
  }
  responseApi(res) {
    console.log('token:', res.accessToken)
  }
  checkLoginState() {
    this.status(this.getStatus.bind(this))
  };
  loginFacebook() {
    this.login(this.getStatus.bind(this))
  }
  logoutFacebook() {
    this.logout()
  }
  render() {
    return (
      <div>
        <button onClick={ this.checkLoginState.bind(this) }>Get Facebook Login Status</button>
        <button onClick={ this.loginFacebook.bind(this) }>Facebook Login</button>
        <button onClick={ this.logoutFacebook.bind(this) }>Facebook Logout</button>
    );
  }
}

export default LoginHOC(configureLoginProps)(App);
```

# License
MIT


docs/CHANGELOG.md
# CHANGELOG


### v1.1.1

**2016 July 23**

* clear npm publish ignore file

### v1.1.0

**2016 July 23**

* Added Testing
* Added logout method


test/setup.js
import jsdom from 'jsdom';
import sinon from 'sinon';

if (typeof document === 'undefined') {
  global.document = jsdom.jsdom('<!doctype html><html><body><script></script></body></html>');
  global.window = document.defaultView;
  global.navigator = global.window.navigator;
}


test/react-facebook-login-hoc-spec.js
import expect from 'expect'
import TestUtils from 'react-addons-test-utils'
import { shallow } from 'enzyme';
import React, { PropTypes, Component } from 'react'
import LoginHOC from '../src/index'

describe('React', () => {
  describe('LoginHOC', () => {

    const configureLoginProps = {
      scope: 'public_profile',
      xfbml: false,
      cookie: false,
      version: 2.6,
      language: 'en_US',
      appId: '1293827123212'
    }

    @LoginHOC(configureLoginProps)
    class App extends Component {
      constructor(props) {
        super(props);

        this.status = this.props.fb.status
        this.login = this.props.fb.login
      }
      getStatus(response) {
        if (response.authResponse) {
          this.responseApi(response.authResponse);
        }
      }
      checkLoginState() {
        this.status(this.getStatus)
      };
      loginFacebook() {
        this.login(this.getStatus)
      }
      render() {
        return (
          <div>
            <button ref='status' onClick={ this.checkLoginState.bind(this) }>Get Facebook Login Status</button>
            <button ref='login' onClick={ this.loginFacebook.bind(this) }>Facebook Login </button>
          </div>
        );
      }
    }

    it('is React Element', () => {
      expect(TestUtils.isElement(<App />)).toEqual(true)
    })

    it('get correct props', () => {
      const wrapperProps = shallow(<App />).first().node

      expect(wrapperProps.props.scope).toEqual('public_profile')
      expect(wrapperProps.props.xfbml).toEqual(false)
      expect(wrapperProps.props.cookie).toEqual(false)
      expect(wrapperProps.props.version).toEqual(2.6)
      expect(wrapperProps.props.language).toEqual('en_US')
      expect(wrapperProps.props.appId).toEqual('1293827123212')
    })

    it('get corrent extend fb method type of function', () => {
      const wrapperFbProps = shallow(<App />).first().node.props.fb;
      expect(typeof wrapperFbProps).toEqual('object')
      expect(typeof wrapperFbProps.login).toEqual('function')
      expect(typeof wrapperFbProps.logout).toEqual('function')
      expect(typeof wrapperFbProps.status).toEqual('function')
    })
  })
})


src/component/login-hoc.react.js
import React, { PropTypes, Component } from 'react'

const LoginHOC = ({ appId, version, language, xfbml, cookie, scope }) => (WrapperComponent) => {
  return class extends Component {

    static defaultProps = {
      scope: scope || 'public_profile',
      xfbml: xfbml || false,
      cookie: cookie || false,
      version: version || 2.6,
      language: language || 'en_US',
      appId: appId || console.error('AppId Error: you must be have appId')
    }

    static propTypes = {
      scope: PropTypes.string.isRequired,
      appId: PropTypes.string.isRequired,
      xfbml: PropTypes.bool,
      cookie: PropTypes.bool,
      version: PropTypes.number,
      language: PropTypes.string
    }

    constructor(props) {
      super(props);

      this.getLoginStatus = (cb) => this.getStatus(cb)
      this.login = (cb, scope) => this.loginFacebook(cb, scope)
      this.logout = () => this.logoutFacebook()
    }

    getStatus(cb) {
      window.FB.getLoginStatus(cb)
    }

    loginFacebook(cb, scope) {
      window.FB.login(cb, scope)
    }

    logoutFacebook() {
      window.FB.logout()
    }

    appendFbRoot() {
      const fbRoot = document.createElement('div');
      fbRoot.id = 'fb-root';
      document.body.appendChild(fbRoot);
    }

    init() {
      window.fbAsyncInit = () => {
        console.log('init')
        window.FB.init({
          version: `v${version}` || `v${this.props.version}`,
          appId: appId || this.props.appId,
          xfbml: xfbml || this.props.xfbml,
          cookie: cookie || this.props.cookie,
        })
      }
    }

    loadSDK(language) {
      ((d, s, id) => {
        const element = d.getElementsByTagName(s)[0];
        const fjs = element;
        let js = element;
        if (d.getElementById(id)) {return;}
        js = d.createElement(s); js.id = id;
        js.src = `//connect.facebook.net/${language}/all.js`;
        fjs.parentNode.insertBefore(js, fjs);
      })(document, 'script', 'facebook-jssdk');
    }

    componentDidMount() {
      //init SDK
      this.appendFbRoot()
      this.init.call(this)
      this.loadSDK(language)
    }

    render() {
      return (
        <WrapperComponent
          { ...this.props }
          fb={ {
            status: this.getLoginStatus,
            login: this.login,
            logout: this.logout
          } } />
      )
    }
  }
}

export default LoginHOC


example/app.react.js
import React, { Component } from 'react'

import LoginHOC from '../src/index'

const configureLoginProps = {
  scope: 'public_profile',
  xfbml: false,
  cookie: false,
  version: 2.6,
  language: 'en_US',
  appId: '488387194689361'
}

class App extends Component {
  constructor(props) {
    super(props)

    this.status = this.props.fb.status
    this.login = this.props.fb.login
    this.logout = this.props.fb.logout
  }
  getStatus(response) {
    if (response.authResponse) {
      this.responseApi.call(this, response.authResponse)
    }
  }
  responseApi(res) {
    console.log('token:', res.accessToken)
  }
  checkLoginState() {
    this.status(this.getStatus.bind(this))
  };
  loginFacebook() {
    this.login(this.getStatus.bind(this))
  }
  logoutFacebook() {
    this.logout()
  }
  render() {
    return (
      <div>
        <button onClick={ this.checkLoginState.bind(this) }>Get Facebook Login Status</button>
        <button onClick={ this.loginFacebook.bind(this) }>Facebook Login</button>
        <button onClick={ this.logoutFacebook.bind(this) }>Facebook Logout</button>
      </div>
    );
  }
}

export default LoginHOC(configureLoginProps)(App);
